Start clearing up annotation syntax for functions. (#3617)
modelica/ModelicaSpecification: chapters/functions.tex

\section{Function Inlining and Event Generation}\label{function-inlining-and-event-generation}

The annotations listed below affect inlining of functions and the related topic of event generation inside functions.

> For the semantics of the annotations see \cref{notation-for-annotation-definitions}.

Reviewer: Now that #3621 is introducing sections more specifically on the semantics, I don't think we should call it _semantics_ and link to a section on the notation.
```suggestion
See \cref{notation-for-annotation-definitions} regarding the notation used to describe the annotations.
```

If {\lstinline!Inline = false!}, the model developer proposes to not inline the function.

> The default for inlining is tool-specific.

Reviewer: Would probably look nicer if combined with the previous sentence to form a two-sentence paragraph.

\begin{nonnormative}
{\lstinline!Inline = true!} is for example used in {\lstinline!Modelica.Mechanics.MultiBody.Frames!} and in functions of {\lstinline!Modelica.Media!} to have no overhead for function calls such as resolving a vector in a different coordinate system and at the same time the function can be analytically differentiated, e.g., for index reduction needed for mechanical systems.
\end{nonnormative}
\end{semantics}
\end{annotationdefinition}

\begin{annotationdefinition}[LateInline]
\begin{synopsis}\begin{lstlisting}

> /*literal*/ constant Boolean LateInline = false;

Reviewer: I thought we decided to avoid writing out defaults like this, as it gives the impression that `LateInline = false` is what one is supposed to write?

Author: As I recall the idea was not use the form when specifying the default literal value (assuming it exists), when it is so "normal" that specifying it would look like an error (like singleInstance).

Different inlining variant and GenerateEvents don't fall under that; and is more similar to absoluteValue.

Reviewer: That's not how I recall it.  For me, the main problem which was resolved when removing all those defaults was that what the reader sees first shouldn't be something one normally doesn't write when using the annotation.  This was avoided by drastically reducing the use of declaration equations for explaining defaults in case of component style annotations.

I find `LateInline` similar to `DocumentationClass`, for which the default was removed in https://github.com/modelica/ModelicaSpecification/pull/3448/commits/aa6ce474bb2c502e465d1ceb270eeba26c56ce1b.

Author: To me these two explanations are fairly similar.
We agree that using `DocumentationClass = false` does not really make sense at all in practice, the point is that `LateInline = false` is just a bit unusual - one can easily see someone experimenting with setting LateInline to true or false to test the result.

And most of the removed defaults in that commit weren't literal values, but explanations as defaults.

Reviewer: OK, so we have different recollections of the exact formulations used in the meeting discussions behind removing literal defaults. Fine.

Now, if we think about how to write clear specification text, don't you agree that there is a problem with picking the style of presentation based on vague opinions about what is more or less unusual or odd?  … that specifying defaults for component style annotations sometimes with declaration equation and sometimes in text (even when the default corresponds to a literal value), is causing more confusion than useful brevity?

Author: But we have lots of annotations with defaults, so introducing a rule saying that we shouldn't use defaults doesn't make sense.

The following are some annotations that are defined with defaults: Documentation.info, Documentation.revisions, Documentation.figures (and many sub-elements), Documentation.styleSheets, experiment.StartTime, and for graphical items origin, rotation, preserveAspectRatio, initalScale, visible, primitivesVisible, fontSize, textColor, color, pattern, ....

Thinking more it might be that we should reconsider the previous idea of not using defaults in some cases.

Reviewer: There's a difference between component style annotations and record style annotations.  As far as I can see, all of your examples are within such records:
- `Documentation`, `experiment`, `GraphicItem`, `CoordinateSystem`, `Text`

The point I am raising in this thread is about component style annotations.

Late inlining is especially useful for differentiation and inversion of functions; for efficiency reasons it is then useful to replace all function calls with identical input arguments by one function call, before the inlining.
\end{nonnormative}

> The default for late inlining is tool-specific, in particular tools may automatically delay inlining to differentiate or possible invert the function before inlining it.

Reviewer: Trying to fix some language:
```suggestion
The default for late inlining is tool-specific.
In particular, tools may automatically delay inlining in order to take advantage of function annotations for derivatives and inverses.
```

Author: ok, handled both of the ones above.

> 

Reviewer: Above, the default was described after giving the meaning of `Inline = false`.  It would make sense to be consistent and describe the default after the meaning of `LateInline = false`.

Reviewer: Replaced by new conversation.  Resolving.

{\lstinline!Inline = false, LateInline = true!} is identical to {\lstinline!LateInline = true!}.

The default for late inlining is tool-specific.
In particular, tools may automatically delay inlining in order to take advantage of function annotations for derivatives and inverses.

> 

Reviewer: I think this would fit even better before the three rules about identical meaning of annotation combinations.

Author: now done

Reviewer: Looks good.  Resolving.

\end{annotationdefinition}

\begin{annotationdefinition}[InlineAfterIndexReduction]
\begin{synopsis}\begin{lstlisting}
/*literal*/ constant Boolean InlineAfterIndexReduction;
\end{lstlisting}\end{synopsis}
\begin{semantics}
Has only an effect within a function declaration.

If {\lstinline!true!}, the model developer proposes to inline the function after the function is differentiated for index reduction, and before any other symbolic transformations are performed.
This annotation cannot be combined with annotations {\lstinline!Inline!} and {\lstinline!LateInline!}.

> The default is to not perform this specific inlining.

Reviewer: How about swapping order with the previous sentence, and then have a paragraph break before the rule about combining annotations?

Author: Should be done now.

Reviewer: Looks fine.  Resolving.
